mypy/constraints.py

```diff
@@ -347,29 +347,33 @@ def visit_instance(self, template: Instance) -> List[Constraint]:
                     template.type.has_base(instance.type.fullname)):
                 mapped = map_instance_to_supertype(template, instance.type)
                 tvars = mapped.type.defn.type_vars
-                for i in range(len(instance.args)):
+                # N.B: We use zip instead of indexing because the lengths might have
+                # mismatches during daemon reprocessing.
+                for tvar, mapped_arg, instance_arg in zip(tvars, mapped.args, instance.args):
```

Comment:
It's kind of unfortunate that we have to do this. Another option might be to make `args` a property and have it adjust the number of args dynamically if the `TypeInfo` has changed. This could have performance impliciations, however.